test/unity/toolkits/drawing_classifier/test_dc_serialization.cxx

```diff
@@ -176,8 +185,211 @@ BOOST_AUTO_TEST_CASE(test_export_coreml) {
+                               drawing_classifier_mock& dc_other) {
+    constexpr auto my_file = "./test_dc_serialization.cxx.save.txt";
+    remove_if_exist(my_file);
+    std::stringstream ss;
+
+    // save impl from the first mock
+    {
+      std::fstream out_file(my_file,
+                            std::ios_base::out | std::ios_base::binary);
+      turi::oarchive oarch(out_file);
+      dc.save_impl(oarch);
+      out_file.close();
+    }
+
+    // load impl to other dc mock instance
+    {
+      std::fstream in_file(my_file, std::ios_base::in | std::ios_base::binary);
+      turi::iarchive iarch(in_file);
+      dc_other.load_version(iarch, dc_other.get_version());
+      in_file.close();
+    }
+
+    // loading params from model spec
+    // saved file will be used to compare with the model spec saved before
+    constexpr auto my_file_loaded = "./test_dc_serialization.cxx.load.txt";
+    remove_if_exist(my_file_loaded);
+
+    {
+      std::fstream out_file(my_file_loaded,
+                            std::ios_base::out | std::ios_base::binary);
+      turi::oarchive oarch(out_file);
+      dc_other.save_impl(oarch);
+      out_file.close();
+    }
+
+    // validation
+    std::ifstream file(my_file,
+                       std::ios::in | std::ios::binary | std::ios::ate);
+    if (!file.is_open()) {
+      std::stringstream ss;
+      ss << "fail to open" << my_file << std::endl;
+      throw std::ios_base::failure(ss.str());
+    }
+    std::streamsize file_size = file.tellg();
+    file.seekg(0, std::ios::beg);
+
+    std::vector<char> model_spec_src(file_size);
+    file.read(model_spec_src.data(), file_size);
+    file.close();
+
+    // open the saved model file and seek the cursor to EOF
+    // in order to tell the size of the file
+    file.open(my_file_loaded, std::ios::in | std::ios::binary | std::ios::ate);
+    if (!file.is_open()) {
+      std::stringstream ss;
+      ss << "fail to open" << my_file_loaded << std::endl;
+      throw std::ios_base::failure(ss.str());
+    }
+    TS_ASSERT_EQUALS(file_size, file.tellg());
+    file_size = file.tellg();
+    // reset the cursor
+    file.seekg(0, std::ios::beg);
+
+    std::vector<char> model_spec_loaded(file_size);
+    file.read(model_spec_loaded.data(), file_size);
+    file.close();
+
+    // compare
+    TS_ASSERT_EQUALS(std::strncmp(model_spec_loaded.data(),
+                                  model_spec_src.data(), file_size),
+                     0);
+
+    // clean myself
+    if (std::remove(my_file)) {
+      std::stringstream ss;
+      ss << "fail to remove file:" << my_file << std::endl;
+      std::perror(ss.str().c_str());
+    }
+    if (std::remove(my_file_loaded)) {
+      std::stringstream ss;
+      ss << "fail to remove file:" << my_file_loaded << std::endl;
+      std::perror(ss.str().c_str());
+    }
+
+    // compare weights in memory
+    auto original_spec = dc.get_model_spec_copy();
```

Comment:
Can we remove the file comparison above and just compare the in-memory representations as here?

Reply:
I think it's also good for us to compare the file written to disk.

Reply:
Why?

Reply:
okay, that's right. we cannot verify whether save works or not by comparing the file.